### PR TITLE
Enable support for LTC Input/Output Mapping

### PR DESCRIPTION
**Pull request by henrytwo**
Complements this upstream PyTorch PR: https://github.com/pytorch/pytorch/pull/75828

This PR adds support for mapping input and output tensors which alias each other. (e.g. maps input weight tensor in parameter to the same tensor in output after a training iteration)
```
MLIR: 
func @graph(%arg0: !torch.vtensor<[1,5],f32>, %arg1: !torch.vtensor<[1],si64>, ..., %arg6: !torch.vtensor<[10,5],f32>, %arg7: !torch.vtensor<[10],f32>, ...) {
  ...
  return %arg0, %arg1, %17, %23, ... : !torch.vtensor<[1,5],f32>, !torch.vtensor<[1],si64>, !torch.vtensor<[10,5],f32>, !torch.vtensor<[10],f32>, ...
}

Input/Output Alias Mapping: 
Output: 0 -> Input: 0
Output: 1 -> Input: 1
Output: 2 -> Input: 6
Output: 3 -> Input: 7
```

The `aten::detach` op has also been disabled in this PR to fix the issue of tensors not aliasing properly due to copying.

cc: @antoniojkim @ke1337

### CODE REVIEW

**Review by silvasean — APPROVED** [review]

**Comment by henrytwo** [comment]
@silvasean Keeping a map of aliased output -> input tensors is a feature that we're planning to use internally at Cerebras -- is this something you'd like us to keep in the generic `TorchMlirLoweringContext` too? Not sure if anyone else would find this useful

**Comment by henrytwo** [comment]
> Isn't that something that upstream already provides or could be moved there?

Upstream we're just calling a method in the `LoweringContext` interface to save the alias, so it's up to the lowering context to figure out what to do with it. Other vendors may want to handle this differently than us (e.g. XLA), so that's why we're implementing it on the `torch-mlir` side.

File: python/torch_mlir/csrc/base_lazy_backend/mlir_lowering_context.h [hunk 1]
```diff
@@ -67,8 +67,12 @@ class TORCH_API TorchMlirLoweringContext : public torch::lazy::LoweringContext {
       const std::vector<int64_t>& param_index,
       bool must_alias = false) override;
 
+  // Check if parameter shape matches result at index.
+  bool CheckResultShape(
+      const BackendDataPtr& parameter_data, int64_t result_idx) override;
+
   // Get the shape of the result tuple component, given by index.
-  c10::optional<torch::lazy::Shape> GetResultShape(size_t index) const override;
+  torch::lazy::Shape GetResultShape(size_t index) const override;
```
Comment (antoniojkim) [inline python/torch_mlir/csrc/base_lazy_backend/mlir_lowering_context.h:75]: is `GetResultShape` needed anymore? can it be merged with `CheckResultShape`?

Comment (henrytwo) [reply python/torch_mlir/csrc/base_lazy_backend/mlir_lowering_context.h:75]: I guess it could, let me check to see if there are any other uses of it.
